### DISCUSSION TITLE
When is `wayland-restore-token` reset on the Wayland side?

### DISCUSSION THREAD

**Discussion by hadim** [rustdesk/rustdesk#10216] [Q&A]
I control a remote Ubuntu 24.10 (xdg-desktop-portal 1.18.4-1ubuntu4) machine running Wayland and occasionally, when trying to connect to it with RustDesk, it failed asking for the user to validate a share screen prompt on the remote machine (which is not ideal ofc).

I know Wayland has been tricky to support in RustDesk (I have been through many past issues already) and I want to thank you since the current support works really well so far (besides the restore issue ofc).

I am wondering if someone knows when this `wayland-restore-token` is being reset on the Wayland side, and so when a new prompt is required on the remote machine? Does that depend on a timeout? a reboot? a screen lock? 

Having that information would be very helpful!

---

As a side note, do you know what would it take to actually improve the current situation? is it something that must be done on the rustdesk side or the wayland side? If someone wants to give a few technical details, maybe that would help a potential future contributor.

**Comment by xlionjuan** [comment]
Always remember, do not use Wayland if you need unattended access.

**Reply by hadim** [reply]
Thank you, but I am already aware of this :-)

My question was about better knowing the inner mechanisms of rustdesk <-> wayland (ref to the original post).

**Reply by xlionjuan** [reply]
What I can sure is Wayland still can't remember the permission after session restart, this is why it can't be unattended access

**Comment by hadim** [comment]
That sounds strange because my second machine (mostly the same as the first one: Ubuntu, NVIDIA card + Wayland) never asked for the screen sharing prompt and so for a few months (at least) allows unattended access.

I am pretty sure this is a config issue on the Ubuntu/Wayland/Gnome side. 

If anyone has an idea for a config I can double-check and diff between those two machines, please let me know.

**Comment by xlionjuan** [comment]
> Ubuntu, NVIDIA card + Wayland) never asked for the screen sharing prompt and so for a few months (at least) allows unattended access.

This should be considered to a bug.

**Comment by fufesou** [comment]
@hadim Hi, you mean the machine can remember the token but it's easy to reset it?

**Reply by hadim** [reply]
What do you mean by "but it's easy to reset it?"?

What I observe is that it's been a while I didn't get a "sharing screen" prompt on that machine.

Do you know the logic that governs the wayland token? which layer is taking care of it? how long for it to be invalidated? does it invalidate on reboot? etc

**Reply by xlionjuan** [reply]
>  how long for it to be invalidated? 

It should be invalid after reboot, you said your computer didn't show up the portal after reboot, it is unexpected behavior.

**Comment by Quackdoc** [comment]
Wayland token is to be reset every time said token gets used, a new token is initialized when the screencast session starts.

XDG portal needs persist_mode=2 which it seems like rustdesk sets, Perhaps the token isn't properly being refreshed, Or perhaps your compositors implementation is... not great. 

Sadly that is common, not every compositor implements the portal the same way and it becomes a very inconsistent experience.

**Reply by hadim** [reply]
Thank you. This is helpful.

So the issue is more on the Wayland compositor side rather than the rustdesk side. Maybe we should poke upstream devs about it? I would be happy to do it but I am not sure which library/logical layer to target? The Ubuntu compositor or some xdg portal package?

**Reply by hadim** [reply]
From https://docs.flatpak.org/en/latest/portal-api-reference.html

> persist_mode u
> How this session should persist. Default is 0. Accepted values are:
> 
> 
> 0: Do not persist (default)
> 1: Permissions persist as long as the application is running
> 2: Permissions persist until explicitly revoked
> 
> Remote desktop screen cast sessions cannot persist. The only allowed persist_mode for remote desktop sessions is 0.
> 
> If the permission for the session to persist is granted, a restore token will be returned via the "Response" signal of the org.freedesktop.portal.ScreenCast.Start method.
> 
> This option was added in version 4 of this interface.

**Reply by Quackdoc** [reply]
no idea, to begin with, I don't know if rustdesk even properly updates the token.

the first thing to do would be to narrow down specifically when rustdesk fails to start, does it
- [ ] fail to start if you have started and stopped a session more then 2 times (Simply try rustdesk a lot back to back for testing). This would likely be a rustdesk issue
- [ ] fail to start if you reboot your PC after sessions (likely a portals issue)
- [ ] fail to start if you run a session once (get the dialog), wait a couple days (can probably force clocks ahead), and try again?

getting to know the issues consistency would narrow down where the issue lies. or code spelunking

**Reply by hadim** [reply]
So I originally got the issue while doing some screen recording on the remote machine using OBS studio. Now I am done with OBS studio, I don't see the screen sharing prompt anymore (unless I reboot or logout/login).

I regularly use that remote machine, so I will keep testing it and report here.
